### ISSUE TITLE
Authenticate users against services

### ISSUE THREAD

**Issue by KellerFuchs** [hashbang/hashbang#7]
Currently, we have services whose user db is completely disconnected from our Unix passwd db:
- [ ] IRC;
- [ ] mail (no sender auth at all, currently!);
- [ ] ... ?

My proposal is to use a GSSAPI implementation that issues (on a shell server) signed tickets that can be used for authenticating against services.
The advantage is that this is generic (useable with anything that supports SASL, for instance), doesn't depend on the availability of a 3rd remote component (i.e. only the shell server and the remote service are involved), and has simple security implications.

**Comment by RyanSquared** [comment]
If we used oidentd to authenticate a user's username on IRC (the portion between the `!` and `@`, in my case, `ryan!ryan@ny1.hashbang.sh`, it would be `ryan`) and enabled reverse DNS lookup of the address connecting to the IRC server (which enables the `ny1.hashbang.sh` in the above example) this solves IRC easily without having to add any extra modules to the IRCd.

**Comment by KellerFuchs** [comment]
Mostly the same comments as on IRC:
- Is it secure?
  - [ ] Is it possible for an attacker to impersonate another user?  
    Yes in the current config: identd usernames get truncated, and somebody creating a `kellerfuch2` account could impersonate me.
  - [ ] Is it possible for an attacker to impersonate or MitM the identd daemon?  
    There is no authentication on that connection, ~~and it is on a port > 1024: any user could try to bind to that port as soon as the server boots (or after identd crashes).~~  
    We currently have no protection against DNS poisoning on the containers.
- [ ] The NickServ account creation is missing.
- It's a solution that only applies to IRC.

EDIT: I confused the RFC number and the port number...

**Comment by RyanSquared** [comment]
I'm still confused on why we need NickServ account creation if we have this solution. Also, as I have said before on IRC, I will be changing the config to allow a username of up to 30 characters. And yes, it's a solution that applies only to IRC, but this would be automated so we wouldn't even need to do any processing.

**Comment by KellerFuchs** [comment]
@ChickenNuggers You didn't answer the comments about whether that's secure.
Also, the maximal username length is 31.

Regarding the NickServ account creation, the whole point of the issue is that we should automatically authenticate/identify shell users, rather than requiring manual user creation and service-specific password management.

> this would be automated so we wouldn't even need to do any processing

No idea what you mean there.

**Comment by RyanSquared** [comment]
> You didn't answer the comments about whether that's secure.

You can't spoof the ident daemon on the shell servers. It's an opt-in in the config and I specifically denied it. It's also on port 113 which users can't host a service on even if they wanted to.

> we should automatically authenticate/identify shell users

They're automatically identified using the ident daemon and you can tell if it's from a #! server by the hostname. I don't see the issue?

**Comment by KellerFuchs** [comment]
> You can't spoof the ident daemon on the shell servers. It's an opt-in in the config

WTF does that even mean.

> I don't see the issue?

Please read again the OP.
The point of this issue is not “being able to tell if users come from #!”, it is “replacing password-based auth for the services we run”.

**Comment by KellerFuchs** [comment]
@ChickenNuggers I'm kinda getting _really_ tired of repeating myself.

> > You can't spoof the ident daemon on the shell servers. It's an opt-in in the config
> 
> WTF does that even mean.

Still not answered.

> Is it possible for an attacker to impersonate or MitM the identd daemon?
> We currently have no protection against DNS poisoning on the containers.

It's funny how you answered the one half of the question which I crossed out and corrected before you even did.

> > > we should automatically authenticate/identify shell users
> > 
> > They're automatically identified using the ident daemon and you can tell if it's from a #! server by the hostname. I don't see the issue?
> 
> Please read again the OP.
> The point of this issue is not “being able to tell if users come from #!”, it is “replacing password-based auth for the services we run”.

Let me restate that yet again, and please fucking read it this time.

Currently, IRC users create NickServ accounts which are handled separately from the shell/mail accounts, and authenticate by password against NickServ.
This issue is about making authentication transparent (and password-less) for shell users on all our services, and you reply with “we can build a thing which is totally unrelated to NickServ”, meaning that it doesn't solve the issue.

**Comment by daurnimator** [comment]
> replacing password-based auth for the services we run

But these would only work when run from the shell servers... right?

If so, @ChickenNuggers's solution seems fine: if a user connects to IRC from a known hashbang shell box ip, we can trust the identd result. Which means we can auth them on that alone.

**Comment by KellerFuchs** [comment]
@daurnimator Yes, we have to support alternate auth methods for people whose client don't run on the shell servers.

> if a user connects to IRC from a known hashbang shell box ip, we can trust the identd result. Which means we can auth them on that alone.

Unfortunately, the NickServ integration story is not as naive, because we cannot know if the IRC account name and the shell account name match. (And if they match, we can't know it's the same user)

**Comment by RyanSquared** [comment]
So I've looked into UnrealIRCd and Anope, and with UnrealIRCd 4 apparently I can use `certfp` as a way to authenticate users with a static certificate in $HOME for IRC.

**Comment by KellerFuchs** [comment]
@RyanSquared Cert-based auth (and I would be much more in favour of using `SASL EXTERNAL`, since SASL gets applied before you are done connecting) might be part of the answer, but as it stands it would be something users can roll out on their own (with manual config), not something which would work transparently.

In particular
- [ ] no account creation, no connection between the shell and IRC account;
- [ ] the cert is manually managed.

**Comment by RyanSquared** [comment]
`certfp` works with certificate and stuff set up on the socket. You really can't get any **before** than that, so `SASL EXTERNAL` would be after `certfp` is taken into effect.

**Comment by KellerFuchs** [comment]
> `certfp` works with certificate and stuff set up on the socket.

I guess you mean it works with a TLS client certificate.  So does `SASL EXTERNAL` (in that context).

> You really can't get any before than that, so `SASL EXTERNAL` would be after `certfp` is taken into effect.

Wrong.

- `certfp` is processed by `NickServ`, *after* you have connected (and potentially even joined channels).
- `SASL EXTERNAL` is processed by the IRCd while you are connecting (and obviously involves cooperation with services, like all the other SASL stuff).

In particular, a recurring issue with `certfp` and other things processed by `NickServ` is that you may fail to jail channels, or not have cloaks applied, or ... because of the race condition between your client and services.

Also, using `SASL` allows us to put restrictions for not-authenticated users, at the IRCd level; for instance, refusing connections from not-authenticated users if we get invaded with SPAM-bots.

**Comment by RyanSquared** [comment]
Alright, I'll look more into a better way of doing this.
